examples/siro_sandbox/sandbox_app.py

```diff
@@ -739,23 +749,26 @@ def _free_camera_lookat_control(self):
         )
 
     def _save_recorded_keyframes_to_file(self):
+        if not self._save_filepath_base or not self._recording_keyframes:
+            return
+
         # Consolidate recorded keyframes into a single json string
         # self._recording_keyframes format:
-        #     List['{"keyframe": {...}', '{"keyframe": {...}', ...]
+        #     List['{"keyframe":{...}', '{"keyframe":{...}',...]
         # Output format:
-        #     '{"keyframes": [{...}, {...}, ...]}'
-        json_keyframes = "".join(
-            keyframe[12:-1] + ","
-            for keyframe in self._recording_keyframes[:-1]
+        #     '{"keyframes":[{...},{...},...]}'
+        json_keyframes = ",".join(
+            keyframe[12:-1] for keyframe in self._recording_keyframes
         )
-        json_keyframes += self._recording_keyframes[-1:][0][
-            12:-1
-        ]  # Last element without trailing comma
-        json_content = '{{"keyframes": [{}]}}'.format(json_keyframes)
+        json_content = '{{"keyframes":[{}]}}'.format(json_keyframes)
 
         # Save keyframes to file
-        with open(self._gfx_replay_save_path, "w") as json_file:
-            json_file.write(json_content)
+        filepath_base = self._find_episode_save_filepath_base()
+        filepath = filepath_base + ".gfx_replay.json.gz"
+        save_as_gzip(json_content.encode("utf-8"), filepath)
+
+        # reset recording_keyframes
+        self._recording_keyframes.clear()
```

Comment:
Gfx-replay keyframes are incremental. Each frame just describes the changes to the visual state since the last keyframe. I.e. Keyframe #2 only makes sense when it immediately follows keyframe #1. We can't arbitrarily clear/discard old ones and then expect a sequence of new ones to "work". We have to be very careful to only do this clear at about the same time that we reset the underlying sim.

I think the solution here is to move this clear to wherever we reset the sim.

Reply:
:white_check_mark:

Reply:
Update here based on more investigation from Ruslan and our Slack chat: it does not work to simply clear old keyframes, even when the sim is reset. If we try this, Ruslan observed that the second gfx-replay file is missing vital `loads`.